cpmpy/solvers/gurobi.py

```diff
@@ -0,0 +1,376 @@
+#!/usr/bin/env python
+"""
+    Interface to the python 'gurobi' package
+
+    Requires that the 'gurobipy' python package is installed:
+
+        $ pip install gurobipy
+    
+    as well as the Gurobi bundled binary packages, downloadable from:
+    https://www.gurobi.com/
+    
+    In contrast to other solvers in this package, Gurobi is not free to use and requires an active licence
+    You can read more about available licences at https://www.gurobi.com/downloads/
+
+    ===============
+    List of classes
+    ===============
+
+    .. autosummary::
+        :nosignatures:
+
+        CPM_gurobi
+
+    ==============
+    Module details
+    ==============
+"""
+
+from .solver_interface import SolverInterface, SolverStatus, ExitStatus
+from ..expressions.core import *
+from ..expressions.variables import _BoolVarImpl, NegBoolView, _IntVarImpl, _NumVarImpl, intvar
+from ..transformations.flatten_model import flatten_constraint, flatten_objective, get_or_make_var
+from ..transformations.get_variables import get_variables
+from ..transformations.linearize import linearize_constraint
+from ..transformations.reification import only_bv_implies
+
+
+class CPM_gurobi(SolverInterface):
+    """
+    Interface to Gurobi's API
+
+    Requires that the 'gurobipy' python package is installed:
+    $ pip install gurobipy
+
+    See detailed installation instructions at:
+    https://support.gurobi.com/hc/en-us/articles/360044290292-How-do-I-install-Gurobi-for-Python-
+
+    Creates the following attributes:
+    user_vars: set(), variables in the original (non-transformed) model,
+                    for reverse mapping the values after `solve()`
+    cpm_status: SolverStatus(), the CPMpy status after a `solve()`
+    tpl_model: object, TEMPLATE's model object
+    _varmap: dict(), maps cpmpy variables to native solver variables
+    """
+
+    @staticmethod
+    def supported():
+        # try to import the package
+        try:
+            import gurobipy as gp
+            from datetime import datetime
+            valid_until = gp.Model().LicenseExpiration
+            today = datetime.today()
+            return today.year * 1e4 + today.month * 1e2 + today.day <= valid_until
+        except ImportError as e:
+            return False
+
+    def __init__(self, cpm_model=None, subsolver=None):
+        """
+        Constructor of the native solver object
+
+        Arguments:
+        - cpm_model: a CPMpy Model()
+        """
+        if not self.supported():
+            raise Exception(
+                "CPM_gurobi: Install the python package 'gurobipy' and make sure your licence is activated!")
+        import gurobipy as gp
+
+        # initialise the native solver object
+        self.env = gp.Env()
+        self.env.setParam("LogToConsole", 0)
+        self.env.setParam("OutputFlag", 0)
+        self.env.start()
+
+        self.gbi_model = gp.Model(env=self.env)
```

Comment:
it seems grb_model rather than gbi would be a more common abbreviation ; )

In this review I'll focus on the gurobi.py file, the linearizations can also be improved but lets do that in a second phase.